Fix trajectories velocity calculation from Velocity mode

LeandroMartinsdS opened the pull request:
This PR fixes the trajectories calculation from velocity mode.

- It considers the buffer rollover for the velocity mode needs information (position or time) from the previous or next buffer.
- Applies fixes in indexing for ppmac startAxisPointsCmd.
- Fixes typo M-variable definition in powerpmac motion program.
- Change the definition of the velocities to Q-variables in motion program.
- Fixes BufferAdr_B in ppmac motion

LeandroMartinsdS committed:
Fix trajectory_scan_code_ppmac.pmc

Remove call for removed subroutine (N112)
Remove assignment of removed variable (VelMode)

Rename clashing variables for ppmac trajectory

Global variable Time clashes with Sys.Time element
User defined variable Status breaks the Status screen IDE

LeandroMartinsdS committed:
Fix axes indexing for geobrick trajectory

Make the necessary changes to tpmac startAxisPointCmd to keep the same
functioning

LeandroMartinsdS committed:
Tidy trajectory definitions

Move velocities to Q variables
Fix Z_Axis definition for trajectory scan

LeandroMartinsdS committed:
Handle buffer rollover

Handle buffer rollover in trajectory velocity calculation from VelMode

Rollback getVelocityMode

JamesOHeaDLS reviewed (commented):
Meant to place in-line comment

JamesOHeaDLS commented on pmacApp/pmc/trajectory_scan_code_ppmac.pmc:
```diff
@@ -6,7 +6,7 @@
 BufferLength = BuffLen                          // BuffLen defined in header file
 BufferAdr_A = 0                                 // BufferAdr defined in header file
 BufferAdr_B = 2 * BufferLength                  // each individual array holds 2 buffers
-Status = 0
```
Suggestion:

Create an enum for the Traj_Status via #defines at the top of the file, to make the checking and setting of this variable clearer throughout the code.

eg Instead of Traj_Status = 2, Traj_Status = Idle_State

A similar thing is done within the automated homing routine PLCs

LeandroMartinsdS replied:
Agree, but I would keep comments with the numeric values anyway, because we will have them when debugging with IDE watchtables, plots, or scopes.

LeandroMartinsdS replied:
Will be using keywords from [docs/source/design.rst](https://github.com/DiamondLightSource/pmac/blob/fixTrajectories/docs/source/design.rst), including the prefix Traj_Status

> Status of motion program for EPICS - 0: Initialised, 1: Active, 2: Idle, 3: Error

LeandroMartinsdS replied:
Done https://github.com/DiamondLightSource/pmac/pull/124/commits/d4e7c80e6db3a8e011e4394f72b8a7048667280e

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -2785,6 +2787,19 @@ asynStatus pmacController::initializeProfile(size_t maxPoints) {
+  }
+  tScanPrevBufferPositions = (double **) malloc(sizeof(double) * PMAC_MAX_CS_AXES);
+  for (int axis = 0; axis < PMAC_MAX_CS_AXES; axis++) {
+    tScanPrevBufferPositions[axis] = (double *) malloc(sizeof(double) * 2);
```
Maybe worth a comment here to explain why the * 2

LeandroMartinsdS replied:
Added in ec4285f8b172b0dd4a035d5239fb91066052e311

LeandroMartinsdS committed:
Fix ppmac trajectory buffer length & definitions

LeandroMartinsdS committed:
Reset tScanPendingPoint_ at the beginning of a scan

LeandroMartinsdS committed:
Fix velocities addresses in pmc for turbopmac

A_Vel_Adr was being calculated from the previous Z_Adr, which represents the Z position at the previous iteration.
As the addresses are cascaded, all the velocities were shifted.

Tidy up trajectory_scan_code for turbo PMAC

Move the buffers addresses update to the end of subroutine
making it more readable

JamesOHeaDLS commented on pmacApp/src/pmacTrajectory.cpp:
```diff
@@ -294,6 +294,25 @@ asynStatus pmacTrajectory::getUserMode(int index, int *user) {
 
+asynStatus pmacTrajectory::getVelocityMode(int index, int *velocityMode) {
+  asynStatus status = asynSuccess;
+  static const char *functionName = "readVelocityMode";
```
functionName is set to 'readVelocityMode' instead of 'getVelocityMode'

There are a few functions in this section with the same mismatch

LeandroMartinsdS replied:
Done in 00568ed39924bc5af541a8fb5cf7ea871cc899f4

JamesOHeaDLS commented on pmacApp/src/pmacTrajectory.cpp:
```diff
@@ -294,6 +294,25 @@ asynStatus pmacTrajectory::getUserMode(int index, int *user) {
+
+  debug(DEBUG_TRACE, functionName, "Called with index", index);
+
+  // Check the index is valid
+  if (index < 0 || index >= noOfValidPoints_) {
+    debug(DEBUG_ERROR, functionName, "Invalid index requested", index);
+    status = asynError;
+  }
```
This index checking code block is used multiple times in this file.  Might be worth making it a function

LeandroMartinsdS replied:
Also making the axis checking into a function.

LeandroMartinsdS replied:
Done in [9dbb012](https://github.com/DiamondLightSource/pmac/pull/124/commits/9dbb012d0c76056907989cc63d24ca97e3250711)

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -2988,6 +3003,23 @@ asynStatus pmacController::buildProfile(int csNo) {
+    } else {
+      debug(DEBUG_ERROR, functionName, "Inconsistent tScanPendingPoint_");
+    }
+  }
```
This block of code appears twice.  Might be worth making it a function

LeandroMartinsdS replied:
Done in [af68554](https://github.com/DiamondLightSource/pmac/pull/124/commits/af685544c11299a0be88cf8bed7530e1b133a47f)

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -3068,7 +3100,32 @@ asynStatus pmacController::appendToProfile() {
+        debug(DEBUG_ERROR,functionName,"Inconsistent tScanPendingPoint_");
+      }
+    }
+    if(tScanPendingPointReady_ > 0) {
+      if(tScanPendingPointReady_ == tScanAxisMask_) {
+        numPointsToBuild++;
+        tScanPendingPointReady_ = 0;
+      } else {
+        debug(DEBUG_ERROR,functionName, "Inconsistent tScanPendingPointReady_");
+      }
+    }
+
```
This section appears to be the only addition from the previous block.  Is is only applicable in this position?

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4532,30 +4589,87 @@ asynStatus pmacController::tScanBuildProfileArray(double *positions, double *vel
+          status = asynError;
+        }
+
+        inverse_deltaTime = 1000000 / (prevBuffTime+times[0]);
```
1000000 is used as a magic number twice in this section.  What does it represent?

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4532,30 +4589,87 @@ asynStatus pmacController::tScanBuildProfileArray(double *positions, double *vel
+        deltaPos = positions[0] - prevBuffPosition;
+        prevBuffVelocity = inverse_deltaTime * deltaPos;
+        tScanPrevBufferVelocity[axis]=prevBuffVelocity;
+      } else if(prevBuffLastVelMode == 4) {
```
prevBuffLastVelMode should be evaluated against an enum to give meaning without needing a comment

LeandroMartinsdS replied:
Implemented in [09e18f9](https://github.com/DiamondLightSource/pmac/pull/124/commits/09e18f91dd8668fa62d861edc2ca56db089db261)

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4566,34 +4680,104 @@ asynStatus pmacController::tScanCalculateVelocityArray(double *positions, double
+        previousPos = tScanPrevBufferPositions[axis][1];
+
+      } else {
+        // Get previous position from last demand Q7x
+        // Set previous velocity as zero
+        sprintf(command, "&%dQ%d", csNum, (71+axis));
+        if (pBroker_->immediateWriteRead(command, reply) == asynSuccess) {
+          sscanf(reply,"%lf",&previousPos);
+          previousVel = 0.0;
+        } else {
+          debug(DEBUG_ERROR, functionName, "Failed to send command", command);
+          status = asynError;
+          break;
```
This code block is used twice in this section.  Maybe worth turning it into a function

LeandroMartinsdS replied:
It should be possible to use the same function for `case 0` as well.

LeandroMartinsdS replied:
Done in [f884372](https://github.com/DiamondLightSource/pmac/pull/124/commits/f884372fd8fc88e73593bfd202c8d77fc9989792)

JamesOHeaDLS commented on pmacApp/src/pmacController.cpp:
```diff
@@ -4566,34 +4680,104 @@ asynStatus pmacController::tScanCalculateVelocityArray(double *positions, double
     // Real Previous -> Current
     case 1:
-    if(times[index] <=  0) {
+      if(index > 0) {
+        // Get previous position and velocity from the current buffer
+        previousPos = positions[index-1];
+        previousVel = velocities[index-1];
+      } else if(pTrajectory_->getNoOfValidPoints()>1) {
+        // Get previous position and velocity from last buffer sent
+        previousPos = tScanPrevBufferPositions[axis][1];
+        previousVel = tScanPrevBufferVelocity[axis];
+      } else {
+        // Get previous position from last demand Q7x
+        // Set previous velocity to zero
+        sprintf(command, "&%dQ%d", csNum, (71+axis));
+        if (pBroker_->immediateWriteRead(command, reply) == asynSuccess) {
+          sscanf(reply,"%lf",&previousPos);
+          previousVel = 0.0;
+        } else {
+          debug(DEBUG_ERROR, functionName, "Failed to send command", command);
+          status = asynError;
+        }
+      }
+      if(times[index] <=  0) {
         debugf(DEBUG_ERROR, functionName, "Invalid time (%d) at point %d",
                times[index],index);
         status = asynError;
         break;
       }
```
This code block is used twice in this section.  Maybe worth turning it into a function

LeandroMartinsdS replied:
Done in [f884372](https://github.com/DiamondLightSource/pmac/pull/124/commits/f884372fd8fc88e73593bfd202c8d77fc9989792)

JamesOHeaDLS commented on pmacApp/pmc/trajectory_scan_definitions_ppmac.pmh:
```diff
@@ -1,8 +1,8 @@
 // *****************************************************************************************
-#define BuffLen       1000    // Length of buffers
-#define DoubleBuffLen 4000
+#define BuffLen        500    // Length of buffers
```
Worth a comment to say why 500, especially as the DoubleBuffLen variable name doesn't fit anymore with the value change

LeandroMartinsdS replied:
The variable name didn't fit before this change, I've just adjusted the values keeping the same ratio, but I agree that is confusing.

I will review that. I suspect that is related to incorrect address definition of the BufferB
```
// Set Initial Values
BufferLength = BuffLen                          // BuffLen defined in header file
BufferAdr_A = 0                                 // BufferAdr defined in header file
BufferAdr_B = 2 * BufferLength                  // each individual array holds 2 buffers
```
https://github.com/DiamondLightSource/pmac/blob/00568ed39924bc5af541a8fb5cf7ea871cc899f4/pmacApp/pmc/trajectory_scan_code_ppmac.pmc#L6C1-L8C89

LeandroMartinsdS replied:
Solved by 33bda474ff6b00281a9722b4a5253bbf236b7cb1

I have confirmed that as it was:
```
#define BuffLen        500    // Length of buffers
#define DoubleBuffLen 2000    // <= 2864 to prevent P-variables/global variables overflow
```
```
BufferAdr_B = 2 * BufferLength
```
only 2000 P-variables were being reserved, but only 1000 were being actually used.
Each one of the buffers was having the last 500 positions skipped.

That was a problem that I have fixed on the PowerPMAC project, and forgot to propagate into the pmac driver files.
Thanks for raising this.

JamesOHeaDLS commented on pmacApp/pmc/trajectory_scan_code.pmc:
```diff
@@ -133,7 +133,7 @@ N102
     End If
     A_Adr = CalculatedBase + BufferLength
-    A_Vel_Adr = Z_Adr + BufferLength
+    A_Vel_Adr = CalculatedBase + 10*BufferLength
```
Where does the 10 come from?

LeandroMartinsdS replied:
Calculated is shifted 10 times from the `CalculatedBase`, being correspondent to the shift of Time/User buffer (both use the same address for the turbopmac) and the 9 axes position buffers.

I will change this part make it more readable, moving the addresses update to the bottom of this subprogram, as it is for the powerpmac.
That way, the  we can keep as
`A_Vel_Adr = Z_Adr + BufferLength`

LeandroMartinsdS committed:
Fix functionName variables in pmacTrajectory methods

LeandroMartinsdS committed:
Refactor buffer rollover handling

Add comments for tScanPrevBufferPositions

Add tScanGetPreviousPoint function

Refactored to make obtaining the previous point for tScan velocity calculation more modular

LeandroMartinsdS committed:
Add enum for velocity mode

LeandroMartinsdS committed:
Refactor index and axis checking in pmacTrajectory

JamesOHeaDLS commented on pmacApp/src/pmacTrajectory.cpp:
```diff
@@ -320,17 +337,9 @@ asynStatus pmacTrajectory::getPosition(int axis, int index, double *position) {
-    debug(DEBUG_ERROR, functionName, "Invalid axis requested", axis);
-    status = asynError;
-  }
+  status = this->isAxisValid(axis);
```
isAxisValid is checked directly after isIndexValid, writing into the same status variable.  This means that any asynError from isIndexValid could be overwritten by an asynSuccess from isAxisValid, and an invalid index would then be passed into profilePositions_

LeandroMartinsdS replied:
Probably we should wrap isIndexValid with a status checking, like other places of the code.

LeandroMartinsdS replied:
I have done what i mentioned on my last message.

JamesOHeaDLS commented on pmacApp/src/pmacTrajectory.cpp:
```diff
@@ -345,17 +354,10 @@ asynStatus pmacTrajectory::getVelocity(int axis, int index, double *velocity) {
-  }
+  status = this->isAxisValid(axis);
+
+  status = this->isIndexValid(index);
```
Same situation here as above

JamesOHeaDLS reviewed (commented):
isAxisValid and isIndexValid need to be write into unique variables, or via another mean so one doesn't mask the error of the other
